common/tools/dev-tool/src/util/printer.ts

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import chalk from "chalk";
+
+const printModes = ["info", "warn", "error", "success", "debug"] as const;
+
+export type Fn<T = void> = (...values: any[]) => T;
+export type ModeMap<T> = { [k in typeof printModes[number]]: T };
+
+/**
+ * The interface that describes the Printer produced by {@link createPrinter}
+ */
+export interface Printer extends ModeMap<Fn> {
+  (...values: any[]): void;
+}
+
+/**
+ * Gets the filename of the calling function
+ */
+function getCaller(): NodeJS.CallSite | undefined {
```

Comment:
this feels pretty magical.. and does it work across Node versions?

Reply:
It should be portable to anything that uses V8 at least. I tested it in Node 8, 10, 12. It's kind of magical but not too much. It just temporarily changes the representation of the Error stack so that you can get the location of the calling function by iterating over it.